shell/platform/linux/fl_key_event_channel.cc

```diff
@@ -0,0 +1,119 @@
+
+static constexpr char kGtkToolkit[] = "gtk";
+static constexpr char kLinuxKeymap[] = "linux";
+
+struct _FlKeyEventChannel {
+  GObject parent_instance;
+
+  FlBasicMessageChannel* channel;
+};
+
+G_DEFINE_TYPE(FlKeyEventChannel, fl_key_event_channel, G_TYPE_OBJECT)
+
+static void fl_key_event_channel_dispose(GObject* object) {
+  FlKeyEventChannel* self = FL_KEY_EVENT_CHANNEL(object);
+
+  g_clear_object(&self->channel);
+
+  G_OBJECT_CLASS(fl_key_event_channel_parent_class)->dispose(object);
+}
+
+static void fl_key_event_channel_class_init(FlKeyEventChannelClass* klass) {
+  G_OBJECT_CLASS(klass)->dispose = fl_key_event_channel_dispose;
+}
+
+static void fl_key_event_channel_init(FlKeyEventChannel* self) {}
+
+FlKeyEventChannel* fl_key_event_channel_new(FlBinaryMessenger* messenger) {
+  g_return_val_if_fail(FL_IS_BINARY_MESSENGER(messenger), nullptr);
+
+  FlKeyEventChannel* self = FL_KEY_EVENT_CHANNEL(
+      g_object_new(fl_key_event_channel_get_type(), nullptr));
+
+  g_autoptr(FlJsonMessageCodec) codec = fl_json_message_codec_new();
+  self->channel = fl_basic_message_channel_new(messenger, kChannelName,
+                                               FL_MESSAGE_CODEC(codec));
+
+  return self;
+}
+
+void fl_key_event_channel_send(FlKeyEventChannel* self,
+                               FlKeyEventType type,
+                               int64_t scan_code,
+                               int64_t key_code,
+                               int64_t modifiers,
+                               int64_t unicode_scarlar_values,
+                               int64_t specified_logical_key,
+                               GCancellable* cancellable,
+                               GAsyncReadyCallback callback,
+                               gpointer user_data) {
+  g_return_if_fail(FL_IS_KEY_EVENT_CHANNEL(self));
+
+  const gchar* type_string;
+  switch (type) {
+    case FL_KEY_EVENT_TYPE_KEYUP:
+      type_string = kTypeValueUp;
+      break;
+    case FL_KEY_EVENT_TYPE_KEYDOWN:
+      type_string = kTypeValueDown;
+      break;
+    default:
+      g_assert_not_reached();
+  }
+
+  g_autoptr(FlValue) message = fl_value_new_map();
+  fl_value_set_string_take(message, kTypeKey, fl_value_new_string(type_string));
+  fl_value_set_string_take(message, kKeymapKey,
+                           fl_value_new_string(kLinuxKeymap));
+  fl_value_set_string_take(message, kScanCodeKey, fl_value_new_int(scan_code));
+  fl_value_set_string_take(message, kToolkitKey,
+                           fl_value_new_string(kGtkToolkit));
+  fl_value_set_string_take(message, kKeyCodeKey, fl_value_new_int(key_code));
+  fl_value_set_string_take(message, kModifiersKey, fl_value_new_int(modifiers));
+  if (unicode_scarlar_values != 0) {
+    fl_value_set_string_take(message, kUnicodeScalarValuesKey,
+                             fl_value_new_int(unicode_scarlar_values));
+  }
+  if (specified_logical_key != 0) {
```

Comment:
This is from the previous code, but since `0` means the value is unset, I'd suggest using `std::optional` or defining a constant (e.g., `kUnsetValue`).